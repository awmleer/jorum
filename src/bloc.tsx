import * as React from 'react'

export type AbstractBloc<T extends Bloc> = { new (...args: any[]): T }

export class Bloc {
  static context:React.Context<any> = null
  static getContext<T extends Bloc>(this: {new():T}): React.Context<T> {
    const ctor: typeof Bloc = this as any
    if (ctor.context === null) {
      ctor.context = React.createContext(null)
    }
    return ctor.context
  }
}
