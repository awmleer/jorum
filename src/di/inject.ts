// export function inject(target: Object, propertyKey: string | symbol, parameterIndex: number) {
//   let existingRequiredParameters: number[] = Reflect.getOwnMetadata(requiredMetadataKey, target, propertyKey) || [];
//   existingRequiredParameters.push(parameterIndex);
//   Reflect.defineMetadata(requiredMetadataKey, existingRequiredParameters, target, propertyKey);
// }