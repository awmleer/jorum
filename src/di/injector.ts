class Injector {
  private readonly injectableMetadataSymbol = Symbol('injectableId')
  private injectables: any = {}
  register(Injectable: any) {
    const symbol = Symbol()
    Reflect.defineMetadata(this.injectableMetadataSymbol, symbol, Injectable)
    this.injectables[symbol] = Injectable
    // this.map[typeof Injectable] = Injectable
  }
  resolve(target: any): any {
    const symbol = Reflect.getMetadata(this.injectableMetadataSymbol, target)
    return this.injectables[symbol]
  }
}

export const injector = new Injector()
