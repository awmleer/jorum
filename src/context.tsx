import * as React from 'react'
import {Bloc} from './bloc'

class BlocMap<T extends Bloc> {
  bloc: T
  constructor(blocClass: { new (): T }) {
    this.bloc = new blocClass()
  }
  prev?: BlocMap<Bloc>
}

// const blocMap = new BlocMap()

// export const BlocMapContext = React.createContext(blocMap)

// export class JorumContext extends React.Component {
//   render() {
//     return (
//       <BlocMapContext.Provider value={blocMap}>
//         {this.props.children}
//       </BlocMapContext.Provider>
//     )
//   }
// }
