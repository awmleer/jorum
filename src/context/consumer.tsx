import * as React from 'react'
import {ConstructorType, Bloc, contextSymbol} from '../bloc'
import {ReactNode} from 'react'

interface Props<T extends Bloc> {
  of: ConstructorType<T>
  children: (bloc: T)=>ReactNode
}

export class Consumer<T extends Bloc> extends React.Component<Props<T>, {}> {
  render() {
    const { of } = this.props
    const Context = Reflect.getMetadata(contextSymbol, of)
    return (
      <Context.Consumer>
        {(value: T) => (
          this.props.children(value)
        )}
      </Context.Consumer>
    )
  }
}

export function useBloc<T>(B: ConstructorType<T>): T {
  const Context = Reflect.getMetadata(contextSymbol, B)
  return React.useContext(Context) as T
}
