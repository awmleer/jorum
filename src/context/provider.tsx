import * as React from 'react'
import {ConstructorType, Bloc, contextSymbol} from '../bloc'
import {FC, useContext, useEffect, useRef} from 'react'
import {injectMetadataKey} from '../inject'
import {Subscribable, Unsubscribable} from 'rxjs'
import {effectsMetadataKey} from '../effect'

export interface ProviderProps<T> {
  of: ConstructorType<T>
  args?: any[]
}

type Props<T> = ProviderProps<T> & {
  children: React.ReactNode
}

class BlocContainer {
  public blocType: ConstructorType<Bloc> = null
  
  private _bloc: Bloc = undefined
  public get bloc() {
    return this._bloc
  }
  public set bloc(newBloc: any) {
    if (newBloc === undefined) return
    if (newBloc === this._bloc) return
    this.cleanUp()
    this._bloc = newBloc
    this.initialize()
  }
  
  private effectSubscriptions: Unsubscribable[] = []
  
  initialize() {
    if (!this._bloc) return
    const effects: string[] = Reflect.getMetadata(effectsMetadataKey, this.blocType.prototype) || []
    for (const effect of effects) {
      const stream$ = (this._bloc as {[key: string]: Subscribable<any>})[effect]
      const subscription = stream$.subscribe(doNothing)
      this.effectSubscriptions.push(subscription)
    }
  }
  
  cleanUp() {
    if (typeof this._bloc !== 'object') return
    for (const subscription of this.effectSubscriptions) {
      subscription.unsubscribe()
    }
    this.effectSubscriptions = []
    if (this._bloc.blocWillDestroy) {
      this._bloc.blocWillDestroy()
    }
  }
  
  hasBloc() {
    return this._bloc !== undefined
  }
}

export const Provider: FC<Props<any>> = function Provider<T>(props: Props<T>) {
  function useInjections(Bloc: ConstructorType<T>, args?: any[]) {
    const injects = Reflect.getMetadata(injectMetadataKey, Bloc) || []
    const paramTypes = Reflect.getMetadata('design:paramtypes', Bloc) || []
    for (let inject of injects) {
      const Context = Reflect.getMetadata(contextSymbol, paramTypes[inject])
      const injection = useContext(Context)
      if (args) {
        args[inject] = injection
      }
    }
  }
  
  const containerRef = useRef(new BlocContainer())
  useInjections(props.of, props.args)
  
  if (containerRef.current.blocType !== props.of) {
    containerRef.current.blocType = props.of
    containerRef.current.bloc = new props.of(...props.args)
  }
  
  if (!containerRef.current.hasBloc()) {
    containerRef.current.bloc = new props.of(...props.args)
  }
  
  useEffect(() => {
    return () => {
      containerRef.current.bloc = null
    }
  }, [])
  
  const Context = Reflect.getMetadata(contextSymbol, props.of)
  return (
    <Context.Provider value={containerRef.current.bloc}>
      {props.children}
    </Context.Provider>
  )
}
Provider.defaultProps = {
  args: []
}


export function withProvider<P, T=any>(providerProps: ((props: P) => ProviderProps<T>) | ProviderProps<T>) {
  return function (C: React.ComponentType<P>): React.ComponentType<P> {
    return function WithProvider(props: P) {
      if (typeof providerProps === 'function') {
        providerProps = providerProps(props)
      }
      return (
        <Provider {...providerProps}>
          <C {...props} />
        </Provider>
      )
    }
  }
}


function doNothing() {}
