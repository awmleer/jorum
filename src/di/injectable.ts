import {injector} from './injector'
// import 'reflect-metadata'

export function injectable<T extends {new(...args:any[]):{}}>(constructor: T) {
  // const decorated = class extends constructor {
  //   newProperty = "new property";
  //   hello = "override";
  // }
  // constructor.name
  injector.register(constructor)
}
