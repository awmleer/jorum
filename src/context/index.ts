import { Provider } from './provider'
import { Consumer } from './consumer'

export { Provider, Consumer }
