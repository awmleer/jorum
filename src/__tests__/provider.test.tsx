import * as TestRenderer from 'react-test-renderer'
import {ProviderFC} from '../context/provider'
import {FC} from 'react'
import {injectable} from '../di/injectable'
import {bloc} from '../bloc'
import * as React from 'react'
import {useBloc} from '..'

@injectable
@bloc
class TestBloc {
  test: string = 'aaa'
}

const TestData: FC = () => {
  const testBloc = useBloc(TestBloc)
  return (
    <div>
      {testBloc.test}
    </div>
  )
}

it('provider initialize', function () {
  const renderer = TestRenderer.create(
    <ProviderFC of={TestBloc}>
      <TestData />
    </ProviderFC>
  )
  expect(renderer.toJSON()).toMatchSnapshot()
})

it('provider with use', function () {
  const testBloc = new TestBloc()
  testBloc.test = 'bbb'
  const renderer = TestRenderer.create(
    <ProviderFC of={TestBloc} use={testBloc}>
      <TestData />
    </ProviderFC>
  )
  expect(renderer.toJSON()).toMatchSnapshot()
});
