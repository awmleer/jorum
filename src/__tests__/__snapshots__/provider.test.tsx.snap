// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provider initialize 1`] = `
<div>
  aaa
</div>
`;

exports[`provider with use 1`] = `
<div>
  bbb
</div>
`;
